framework/Floats.c

```diff
@@ -4,82 +4,114 @@
+static_assert(FLOAT32_NAN_EXPONENT == FLOAT32_EXPONENT_MASK, "Float32::NaNs have all exponent bits set");
+static_assert(FLOAT32_INFINITY_EXPONENT == FLOAT32_EXPONENT_MASK, "Float32::Inf has all exponent bits set");
+static_assert(MASK(FLOAT32_EXPONENT_BITS) == FLOAT32_EXPONENT_MASK, "Float32 exponent mask has different size than the field");
+static_assert(MASK(FLOAT32_MANTISSA_BITS) == FLOAT32_MANTISSA_MASK, "Float32 mantissa mask has different size than the field");
+static_assert(QUIET(FLOAT32) == FLOAT32_MANTISSA_QUIET_NAN_MASK, "Quiet bit is MSB of the mantissa");
+static_assert(FLOAT32_SIGN_BITS + FLOAT32_EXPONENT_BITS + FLOAT32_MANTISSA_BITS == 32, "Bitfields sums to type size");
 
-Float80 new_float80(unsigned sign, uint32_t exponent, unsigned jbit, uint64_t mantissa)
+static_assert(sizeof(Float64) == sizeof(double), "Float64 structure is not of the correct size");
+static_assert(FLOAT64_NAN_EXPONENT == FLOAT64_EXPONENT_MASK, "Float64::NaNs have all exponent bits set");
+static_assert(FLOAT64_INFINITY_EXPONENT == FLOAT64_EXPONENT_MASK, "Float64::Inf has all exponent bits set");
+static_assert(MASK(FLOAT64_EXPONENT_BITS) == FLOAT64_EXPONENT_MASK, "Float64 exponent mask has different size than the field");
+static_assert(MASK(FLOAT64_MANTISSA_BITS) == FLOAT64_MANTISSA_MASK, "Float64 mantissa mask has different size than the field");
+static_assert(QUIET(FLOAT64) == FLOAT64_MANTISSA_QUIET_NAN_MASK, "Quiet bit is MSB of the mantissa");
+static_assert(FLOAT64_SIGN_BITS + FLOAT64_EXPONENT_BITS + FLOAT64_MANTISSA_BITS == 64, "Bitfields sums to type size");
+
+static_assert(sizeof(double) < sizeof(long double), "Compiler does not support long double");
+static_assert(sizeof(Float80) == sizeof(long double), "Float80 structure is not of the correct size");
+static_assert(FLOAT80_NAN_EXPONENT == FLOAT80_EXPONENT_MASK, "All Float80::NaNs have all exponent bits set");
+static_assert(FLOAT80_INFINITY_EXPONENT == FLOAT80_EXPONENT_MASK, "Float80::Inf has all exponent bits set");
+static_assert(MASK(FLOAT80_EXPONENT_BITS) == FLOAT80_EXPONENT_MASK, "Float80 exponent mask has different size than the field");
+static_assert(MASK(FLOAT80_MANTISSA_BITS) == FLOAT80_MANTISSA_MASK, "Float80 mask has different size than mantissa+jbit");
+static_assert(QUIET(FLOAT80) == FLOAT80_MANTISSA_QUIET_NAN_MASK, "Quiet bit is MSB of the mantissa");
+static_assert(FLOAT80_SIGN_BITS + FLOAT80_EXPONENT_BITS + FLOAT80_JBIT_BITS + FLOAT80_MANTISSA_BITS == 80, "Bitfields sums to type size");
+
+/** @} */
+
+// function definitions
+Float16 new_random_float16()
 {
-    Float80 f;
-    f.sign = sign & 1u;
-    f.jbit = jbit & 1u;
-    f.exponent = exponent & FLOAT80_EXPONENT_MASK;
-    f.mantissa = mantissa & FLOAT80_MANTISSA_MASK;
+    Float16 f;
+    f.sign = random32();
+    f.exponent = random32();
+    f.mantissa = set_random_bits(random32() % (FP16_MANTISSA_BITS + 1), FP16_MANTISSA_BITS);
+
```

Comment:
nitpick: you're using 96 bits of random to fill a 16-bit value. Due to internals of how `random32()` works, you're actually doing **six** random number calculations to get here. Please use one single `random()` call, which gives you 31 bits of information

Reply:
Thiago, do you mean something like this:
```
auto rand = random32()
f.sign = (ran & 1)
f.exponent = rand >> 1
f.mantissa = set_random_bits(rand >> 17 % (FP16_MANTISSA_BITS + 1), FP16_MANTISSA_BITS);
```

Reply:
Using `random`, not `random32`, but yes.